### PR TITLE
Add transformers to models

### PR DESCRIPTION
**Pull request by BellaT**
- Added a BaseTransformer
- Transformer helper functions
- A transformer for each model with functions for any relationships
- Updated interfaces to be interfaces instead of classes

### CODE REVIEW

File: src/BaseTransformer.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * Perform the data mapping - this should be overridden by each specific Transformer
+     *
+     * @param data
+     * @param args
+     */
+    mapData(data: object, args: Array<any> = []): object {
+        return {
+            data,
+            ...args,
```
Comment (Jamiewarb) [inline src/BaseTransformer.ts]: Not certain we want to return args here. 

Essentially this is just an identity function so it should just return the data as is (as there's nothing to map it to)

File: src/BaseTransformer.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    /**
+     * Transform an attribute if in includes, and the data exists
+     *
+     * @param data
+     * @param includeName
+     * @param transformer
+     */
+    item(data: Record<string, any>, includeName: string, transformer: BaseTransformer) {
```
Comment (Jamiewarb) [inline src/BaseTransformer.ts]: Should we declare the return type of this method as well?

File: src/Interfaces/order/Order.ts [hunk 3]
```diff
@@ -41,9 +41,9 @@ export default class Order {
     placedAt?: Date;
     meta?: string;
     orderStatus?: number;
-    resourceType = 'orders';
+    resourceType?: string;
     // belongs to
-    user?: User;
+    user?: User | null;
```
Comment (Jamiewarb) [inline src/Interfaces/order/Order.ts]: Does the question mark cover the null already? Or is this required?

File: src/Models/order/Transformers/OrderTransformer.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            shippingCountry: order.shippingCountry,
+            shippingZip: order.shippingZip,
+            contactEmail: order.contactEmail,
+            contactPhone: order.contactPhone,
+            meta: order.meta,
+            resourceType: order.resourceType,
+            // belongs to
+            user: this.includeUser(order),
+            // has many
+            lines: this.includeOrderLines(order.lines || []),
```
Comment (Jamiewarb) [inline src/Models/order/Transformers/OrderTransformer.ts]: Shouldn't this be
```suggestion
            lines: this.includeOrderLines(order),
```

File: src/Models/product/Transformers/ProductTransformer.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+            popularity: product.popularity,
+            description: product.description,
+            recentRank: product.recentRank,
+            manufacturer: product.manufacturer,
+            metaKeywords: product.metaKeywords,
+            recommendedAge: product.recommendedAge,
+            metaDescription: product.metaDescription,
+            shortDescription: product.shortDescription,
+            // has many
+            variants: this.includeProductVariant(product.variants || []),
```
Comment (Jamiewarb) [inline src/Models/product/Transformers/ProductTransformer.ts]: Shouldn't this be
```suggestion
            variants: this.includeProductVariant(product),
```

File: src/Models/stock/Transformers/StockTransformer.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        return new Stock({
+            id: stock.id,
+            createdAt: stock.createdAt,
+            updatedAt: stock.updatedAt,
+            statusId: stock.statusId,
+            // belongs to
+            productVariant: this.includeProductVariant(stock),
+            user: this.includeUser(stock),
+            // has many
+            returnOrderLines: this.includeReturnOrderLines(stock.returnOrderLines || []),
```
Comment (Jamiewarb) [inline src/Models/stock/Transformers/StockTransformer.ts]: Shouldn't this be 
```suggestion
            returnOrderLines: this.includeReturnOrderLines(stock),
```

File: src/Models/user/Transformers/UserTransformer.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+            registrationStatusId: user.registrationStatusId,
+            stripeId: user.stripeId,
+            residualChargesOptIn: user.residualChargesOptIn,
+            subscriptionCancellationTerms: user.subscriptionCancellationTerms,
+            subscriptionCancellationCharges: user.subscriptionCancellationCharges,
+            verifiedAt: user.verifiedAt,
+            orderStatus: user.orderStatus,
+            resourceType: user.resourceType,
+            // has many
+            orders: this.includeOrdersCollection(user.orders || []),
```
Comment (Jamiewarb) [inline src/Models/user/Transformers/UserTransformer.ts]:
```suggestion
            orders: this.includeOrdersCollection(user),
```

File: src/Models/order/Transformers/OrderTransformer.ts (hunk 4 above)
Comment (BellaT) [reply src/Models/order/Transformers/OrderTransformer.ts]: It breaks

File: src/Models/product/Transformers/ProductTransformer.ts [hunk 8]
```diff
@@ -29,11 +29,11 @@ export default class ProductTransformer extends BaseTransformer {
             metaDescription: product.metaDescription,
             shortDescription: product.shortDescription,
             // has many
-            variants: this.includeProductVariant(product.variants || []),
+            variants: this.includeProductVariant(product),
         });
     }
 
-    includeProductVariant(product: ProductInterface[]): ProductVariantInterface[] {
+    includeProductVariant(product: ProductInterface): ProductVariantInterface[] {
```
Comment (Jamiewarb) [inline src/Models/product/Transformers/ProductTransformer.ts]: I think this method name should be plural
```suggestion
    includeProductVariants(product: ProductInterface): ProductVariantInterface[] {
```

File: src/Models/user/Transformers/UserTransformer.ts [hunk 9]
```diff
@@ -28,11 +28,11 @@ export default class UserTransformer extends BaseTransformer {
             orderStatus: user.orderStatus,
             resourceType: user.resourceType,
             // has many
-            orders: this.includeOrdersCollection(user.orders || []),
+            orders: this.includeOrdersCollection(user),
         });
     }
 
-    includeOrdersCollection(user: UserInterface[]): OrderInterface[] {
+    includeOrdersCollection(user: UserInterface): OrderInterface[] {
```
Comment (Jamiewarb) [inline src/Models/user/Transformers/UserTransformer.ts]: Following on from the rest of the naming conventions, this should maybe be `includeOrders`
```suggestion
    includeOrders(user: UserInterface): OrderInterface[] {
```

**Pull request merged**